ouroboros-consensus/test-consensus/Test/Dynamic/Util/TestConfig.hs

```diff
@@ -0,0 +1,482 @@
+            Just (Max (CoreNodeId i)) -> i >= n
+            Nothing                   -> False
+        maxLeader =
+            (foldMap . foldMap) (Just . Max) $
+            getLeaderSchedule schedule
+
+{-------------------------------------------------------------------------------
+  MockPBFT test
+-------------------------------------------------------------------------------}
+
+type BlockMockPBFT = SimplePBftBlock SimpleMockCrypto PBftMockCrypto
+
+instance TestableBlock BlockMockPBFT where
+    type AdhocTestConfig BlockMockPBFT = SecurityParam
+
+    genTestConfig = do
+        let k = SecurityParam 5
+
+        numCoreNodes <- arbitrary
+        numSlots     <- arbitrary
+
+        defaultGenTestConfig k numCoreNodes numSlots
+
+    shrinkTestConfig = shrinkTestConfigFreely
+
+    adhocInconsistenciesTestConfig _ = []
+
+{-------------------------------------------------------------------------------
+  RealPBFT test
+-------------------------------------------------------------------------------}
+
+type BlockRealPBFT = ByronBlock
+
+defaultRealPBftParams :: NumCoreNodes -> PBftParams
+defaultRealPBftParams numCoreNodes = PBftParams
+  { pbftNumNodes           = n
+  , pbftSecurityParam      = SecurityParam k
+  , pbftSignatureThreshold = (1 / n) + (1 / k)
+    -- crucially: @floor (k * t) >= ceil (k / n)@
+  }
+    where
+      n :: Num a => a
+      n = fromIntegral x
+        where
+          NumCoreNodes x = numCoreNodes
+
+      k :: Num a => a
+      k = 10
+
+genSlot :: SlotNo -> SlotNo -> Gen SlotNo
+genSlot lo hi = SlotNo <$> choose (unSlotNo lo, unSlotNo hi)
+
+-- | As 'genNodeJoinPlan', but ensures an additional invariant
+--
+-- INVARIANT Under ideal circumstances, this 'NodeJoinPlan' ensures that *the*
+-- chain includes at least @k@ blocks within every @2k@-slot window.
+--
+genRealPBFTNodeJoinPlan :: PBftParams -> NumSlots -> Gen NodeJoinPlan
+genRealPBFTNodeJoinPlan params numSlots@(NumSlots t)
+  | n < 0 || t < 1 = error $ "Cannot generate RealPBFT NodeJoinPlan: "
+    ++ show (params, numSlots)
+  | otherwise      = go (NodeJoinPlan Map.empty) Ref.RealPBFT.emptyState
+  where
+    PBftParams{pbftNumNodes} = params
+    n                        = fromIntegral pbftNumNodes
+
+    lastSlot = SlotNo $ fromIntegral $ t - 1
+
+    go ::
+         NodeJoinPlan
+         -- ^ an /incomplete/ and /viable/ node join plain
+      -> Ref.RealPBFT.State
+         -- ^ a state whose 'Ref.RealPBFT.nextSlot' is <= the last join slot in
+         -- given plan (or 0 if the plan is empty)
+      -> Gen NodeJoinPlan
+    go nodeJoinPlan@(NodeJoinPlan m) st
+      | i == n    = pure $ NodeJoinPlan m
+      | otherwise = do
+            -- Does this slot make the invariant unsatisfiable?
+        let check s' =
+                Ref.RealPBFT.viable params lastSlot
+                    (NodeJoinPlan (Map.insert nid s' m))
+                    st
+        s' <- genSlot (Ref.RealPBFT.nextSlot st) lastSlot
+            `suchThat` check
```

Comment:
I have recently been bitten by `suchThat`: in some rare cases, it was impossible to generate values that satisfied the condition, resulting in hangs in my testsuite. I only noticed it after running > 10k tests. Debugging it was very annoying, mostly because I excluded the possibility that my use of `suchThat` was the cause, as I couldn't conceive of the rare case that triggered the infinite loop.
</war story>

Can you convince yourself that this will never result in an infinite loop? Moreover, will it not needlessly make the testsuite slower (the RealPBFT protocol tests are already slow enough, you probably suffer the most from this already)?

Reply:
Yes, I have suffered by the same thing while working on this test suite! `suchThat` is certainly a double-edged sword. It's very high on my list of suspects for "why is this test hanging?"

So I've been careful in this use of itr, yes. I'll try to elaborate in the comments. I've also been considering a `suchThatNTimes` combinator derived from `Quickcheck`'s `suchThatMaybe`.

Reply:
I realized the `check` predicate is monotonic. So we can use a binary search instead of `suchThat`: see the `test-consensus: replace suchThat with a binary search` commit.